optuna/study/study.py

```diff
@@ -928,6 +945,31 @@ def _pop_waiting_trial_id(self) -> Optional[int]:
 
         return None
 
+    def _should_skip_enqueue(self, params: Dict[str, Any]) -> bool:
+
+        for trial in self.get_trials(deepcopy=False):
+            trial_params = trial.system_attrs.get("fixed_params", trial.params)
+            if trial_params.keys() != params.keys():
+                # Can't have repeated trials if different params are suggested.
+                continue
+
+            for param_name, param_value in params.items():
+                existing_param = trial_params[param_name]
+                if not isinstance(param_value, type(existing_param)):
+                    # Enqueued param has distribution that does not match existing param
+                    # (e.g. trying to enqueue categorical to float param).
+                    # We are not doing anything about it here, since sanitization should
+                    # be handled regardless if `skip_if_exists` is `True`.
+                    continue
+
+                if isinstance(param_value, (float, np.floating)):
```

Comment:
In https://github.com/optuna/optuna/pull/3567/files#r879072605, we use `isinstance(..., Real)` to handle float-like types like `np.floating`. Can we use `numbers.Real` here as well?

Reply:
I'm afraid `mypy` is not happy with that change - `isclose` accepts just about any type, but not `Real`. :sweat_smile:

Reply:
Yes, you are correct. In #3567, `Real` values are casted to `float` to call `np.isnan()`.

https://github.com/optuna/optuna/blob/b213753a816689a97b6433e1e780ca32c4293e07/optuna/distributions.py#L464-L465

Do you think it make sense to cast values with `float()` like:

```python
is_repeated = (
    np.isnan(float(param_value)) or np.isclose(float(param_value), float(existing_param), atol=0.0)
    if isinstance(param_value, Real)
    else param_value == existing_param
)
```

Reply:
Sounds reasonable. Implemented in 87e3c27183533447a0de3561715db62d29d84526.